glean-core/src/event_database/mod.rs

```diff
@@ -0,0 +1,325 @@
+            .and_then(|mut file| writeln!(file, "{}", event_json))
+        {
+            log::error!("Error writing event to store {}: {}", store_name, err);
+        }
+    }
+
+    /// Get a snapshot of the stored event data as a JsonValue.
+    ///
+    /// # Arguments
+    ///
+    /// * `store_name` - The name of the desired store.
+    /// * `clear_store` - Whether to clear the store afterward.
+    ///
+    /// # Returns
+    ///
+    /// The an array of events, JSON encoded, if any.
+    pub fn snapshot_as_json(&self, store_name: &str, clear_store: bool) -> Option<JsonValue> {
+        let result = {
+            let db = self.event_stores.read().unwrap();
+            db.get(&store_name.to_string()).and_then(|store| {
+                if !store.is_empty() {
+                    let first_timestamp = store[0].timestamp;
+                    Some(json!(store
+                        .iter()
+                        .map(|e| e.serialize_relative(first_timestamp))
+                        .collect::<Vec<JsonValue>>()))
+                } else {
+                    None
+                }
+            })
+        };
+
+        if clear_store {
+            {
+                let mut db = self.event_stores.write().unwrap();
+                db.remove(&store_name.to_string());
+            }
+
+            {
+                let _lock = self.file_lock.write().unwrap();
+                if let Err(err) = fs::remove_file(self.path.join(store_name)) {
+                    log::error!("Error removing events queue file {}: {}", store_name, err);
+                }
+            }
+        }
+
+        result
+    }
+
+    /// **Test-only API (exported for FFI purposes).**
+    ///
+    /// Return whether there are any events currently stored for the given even
+    /// metric.
+    ///
+    /// This doesn't clear the stored value.
+    pub fn test_has_value<'a>(&'a self, meta: &'a CommonMetricData, store_name: &str) -> bool {
+        self.event_stores
+            .read()
+            .unwrap()
+            .get(&store_name.to_string())
+            .into_iter()
+            .flatten()
+            .any(|event| event.name == meta.name && event.category == meta.category)
+    }
+
+    /// **Test-only API (exported for FFI purposes).**
+    ///
+    /// Get the vector of currently stored events for the given event metric in
+    /// the given store.
+    ///
+    /// This doesn't clear the stored value.
+    pub fn test_get_value<'a>(
+        &'a self,
+        meta: &'a CommonMetricData,
+        store_name: &str,
+    ) -> Option<Vec<RecordedEventData>> {
+        let value: Vec<RecordedEventData> = self
+            .event_stores
+            .read()
+            .unwrap()
+            .get(&store_name.to_string())
+            .into_iter()
+            .flatten()
+            .filter(|event| event.name == meta.name && event.category == meta.category)
+            .cloned()
+            .collect();
```

Comment:
Chain it ALL!
(Not saying this is wrong or anything, it's just weird seeing such a long call chain :D)

Reply:
I feel that's like most of the Rust code I see in the wild, but again, I'm not sure if it's a good or a bad thing.  If it really seems unmanageable, let me know.  I guess the alternative is something really deeply nested so :man_shrugging: ?

Reply:
All good!